### PR TITLE
KAFKA-4750: RocksDBStore always deletes null values

### PR DESCRIPTION
**Pull request by evis**
@guozhangwang @mjsax

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5696/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5682/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5697/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5683/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    private RocksDBStore<String, String> subject;
+    private MockProcessorContext context;
+
+    @Before
+    public void setUp() throws Exception {
+        Serializer<String> serializer = new StringSerializer() {
+            @Override
+            public byte[] serialize(String topic, String data) {
+                if (data == null)
+                    return "not-null".getBytes();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: I think `"not-null"` might be confusion. I think a better naming would be `"null-encoding-that-is-not-just-'null'"`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                return super.serialize(topic, data);
+            }
+        };
+        Serde<String> serde = Serdes.serdeFrom(serializer, new StringDeserializer());
+        subject = new RocksDBStore<>("test", serde, serde);
+        File dir = TestUtils.tempDirectory();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: `dir` only used once -- can be removed

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        context = new MockProcessorContext(dir,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: NIt: fix indention -- either move `dir` to next line, or align other parameters with `dir` starting column.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    public void setUp() throws Exception {
+        Serializer<String> serializer = new StringSerializer() {
+            @Override
+            public byte[] serialize(String topic, String data) {
+                if (data == null)
+                    return "not-null".getBytes();
+                return super.serialize(topic, data);
+            }
+        };
+        Serde<String> serde = Serdes.serdeFrom(serializer, new StringDeserializer());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: can be `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ * Test checks RocksDBStore behaviour with serializer, which
+ * serializes null value into non-null byte array.
+ */
+public class RocksDBStoreCustomNullSerdeTest {
+    private RocksDBStore<String, String> subject;
+    private MockProcessorContext context;
+
+    @Before
+    public void setUp() throws Exception {
+        Serializer<String> serializer = new StringSerializer() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: can be `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            @Override
+            public byte[] serialize(String topic, String data) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: both parameters can be `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        subject.close();
+    }
+
+    @Test
+    public void shouldNotReturnDeletedInIterator() {
+        subject.init(context, subject);
+        subject.put("a", "1");
+        subject.put("b", "2");
+        subject.delete("a");
+        KeyValueIterator<String, String> it = subject.all();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: can be `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        while (it.hasNext()) {
+            KeyValue<String, String> next = it.next();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: can be `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            if (next.key.equals("a"))
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreCustomNullSerdeTest.java]: Nit: Please use `{ }` (even for one line blocks)

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the patch and the test!

Just a couple of nits :)

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5700/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5714/
Test PASSed (JDK 7 and Scala 2.11).

**Review by mjsax — APPROVED** [review]
LGTM.

Call for review and merging @guozhangwang @dguy

**Comment by guozhangwang** [comment]
@evis @mjsax I have some more thoughts commented on the ticket itself: I'm wondering if we should just not using `put(key, null)` to implement `delete(key)` since our java doc doe not enforce that a `null` value (note this value is null object, not underlying null byte arrays) indicates deletion.

**Comment by guozhangwang** [comment]
I'm closing this PR and continue working on https://github.com/apache/kafka/pull/4508.

**Pull request closed without merging**